### PR TITLE
Implement `view_formats` for `SurfaceConfiguration`

### PR DESCRIPTION
**Pull request by jinleili**
**Checklist**

- [x] Run `cargo clippy`.
- [x] Run `RUSTFLAGS=--cfg=web_sys_unstable_apis cargo clippy --target wasm32-unknown-unknown` if applicable.
- [x] Add change to CHANGELOG.md. See simple instructions inside file.

**Connections**
Follow up to https://github.com/gfx-rs/wgpu/pull/3237

**Testing**
Tested locally except deno part

### CODE REVIEW

**Review by crowlKats — COMMENTED** [review]
Deno part looks good to me.

If you want, i can test the deno change, but am unsure how viewformats are used. Though, given its a simple & straight-forward change, testing it doesnt seem too necessary

**Comment by jinleili** [comment]
Looking at the #3399 implementation, I'm not sure if the Vulkan backend needs to be handled separately here as well.

**Comment by teoxoy** [comment]
Thanks for updating the validation but there is one more rule:
> [Validate texture format required features](https://gpuweb.github.io/gpuweb/#abstract-opdef-validate-texture-format-required-features) of each element of configuration.[viewFormats](https://gpuweb.github.io/gpuweb/#dom-gputexturedescriptor-viewformats) with device.[[[device]]](https://gpuweb.github.io/gpuweb/#dom-gpuobjectbase-device-slot).

Ideally, we'd share some validation for the `TextureDescriptor` and the `configure` fn... but not a blocker.

> Looking at the https://github.com/gfx-rs/wgpu/pull/3399 implementation, I'm not sure if the Vulkan backend needs to be handled separately here as well.

I don't think we need to do anything extra for Vulkan here. @cwfitzgerald can maybe confirm?

**Comment by cwfitzgerald** [comment]
Yeah, I think we need mutable format on the swapchain via https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkSwapchainCreateFlagBitsKHR.html and VK_KHR_swapchain_mutable_format. This is fine on desktop, but has zero support on android.

**Comment by jinleili** [comment]
> Yeah, I think we need mutable format on the swapchain via https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkSwapchainCreateFlagBitsKHR.html and VK_KHR_swapchain_mutable_format. This is fine on desktop, but has zero support on android.

The addition of the corresponding extension `KhrImageFormatListFn` was in #3412

**Comment by teoxoy** [comment]
`VK_KHR_image_format_list` can be optional but it seems to be required if we require `VK_KHR_swapchain_mutable_format`.
I think we should land https://github.com/gfx-rs/wgpu/pull/3412 first and make it required in this PR.

**Comment by jinleili** [comment]
I checked the `VK_KHR_swapchain_mutable_format` extension on Android 13 devices, and sure enough, it's not supported on Android devices.

> VK_KHR_image_format_list can be optional but it seems to be required if we require VK_KHR_swapchain_mutable_format?

Yes, after re-reading the documentation, they should all be required. Since `VK_KHR_imageless_framebuffer` also requires `VK_KHR_image_format_list`, the permission of this extension has to be determined separately.

File: wgpu-hal/src/vulkan/adapter.rs [hunk 1]
```diff
@@ -643,6 +643,15 @@ impl PhysicalDeviceCapabilities {
...
         }
 
+        if self.effective_api_version >= vk::API_VERSION_1_2
+            || self.supports_extension(vk::KhrImageFormatListFn::name())
+        {
+            // Optional `VK_KHR_swapchain_mutable_format`
+            if self.supports_extension(vk::KhrSwapchainMutableFormatFn::name()) {
+                extensions.push(vk::KhrSwapchainMutableFormatFn::name());
+            }
+        }
```
Comment (teoxoy) [inline wgpu-hal/src/vulkan/adapter.rs:653]:
```suggestion
        // Optional `VK_KHR_swapchain_mutable_format`
        if self.supports_extension(vk::KhrSwapchainMutableFormatFn::name()) {
            extensions.push(vk::KhrSwapchainMutableFormatFn::name());
        }
```

I'm quite certain that `supports_extension` returns `false` if the required extensions for this one are not supported.

File: wgpu-hal/src/vulkan/device.rs [hunk 2]
```diff
@@ -565,6 +583,12 @@ impl super::Device {
             .clipped(true)
             .old_swapchain(old_swapchain);
 
+        let mut format_list_info = vk::ImageFormatListCreateInfo::builder();
+        if !raw_view_formats.is_empty() {
+            format_list_info = format_list_info.view_formats(&raw_view_formats);
+            info = info.push_next(&mut format_list_info);
+        }
```
Comment (teoxoy) [inline wgpu-hal/src/vulkan/device.rs:590]: Can we move the builder inside the if?

File: wgpu-hal/src/vulkan/device.rs [hunk 3]
```diff
@@ -546,11 +546,29 @@ impl super::Device {
...
-        let info = vk::SwapchainCreateInfoKHR::builder()
-            .flags(vk::SwapchainCreateFlagsKHR::empty())
+
+        let original_format = self.shared.private_caps.map_texture_format(config.format);
+        let mut raw_flags = vk::SwapchainCreateFlagsKHR::empty();
+        let mut raw_view_formats: Vec<vk::Format> = vec![];
+        if !config.view_formats.is_empty()
+            && self
+                .enabled_device_extensions()
+                .contains(&vk::KhrSwapchainMutableFormatFn::name())
```
Comment (teoxoy) [inline wgpu-hal/src/vulkan/device.rs:556]: This looks like we'd silently ignore view formats.
Considering the extension has almost 0 support on android, we don't want to require it either.

I think we need a new downlevel flag for this capability.

File: wgpu-hal/src/vulkan/device.rs (hunk 2 above)
Comment (jinleili) [reply wgpu-hal/src/vulkan/device.rs:590]: Moving into `if` causes `format_list_info` to have a shorter lifetime than `info`

Comment (teoxoy) [reply wgpu-hal/src/vulkan/device.rs:590]: Hmm, I see.

File: wgpu-hal/src/vulkan/device.rs (hunk 3 above)
Comment (jinleili) [reply wgpu-hal/src/vulkan/device.rs:556]: Now there is a problem I don't know how to handle properly: we already have a `VIEW_FORMATS` downlevel flag, and inside the `frame.texture.create_view()` we no longer know if the texture is from `SurfaceTexture` or not.
https://github.com/gfx-rs/wgpu/blob/964c94a02dd13fdde4faf8ffb4a1b6756aa84fd7/wgpu-core/src/device/mod.rs#L1031-L1037

Comment (teoxoy) [reply wgpu-hal/src/vulkan/device.rs:556]: I think as long as we require the new downlevel flag in the configure fn then we know it should not be possible to get a surface texture with view formats set on it.

Comment (jinleili) [reply wgpu-hal/src/vulkan/device.rs:556]: It looks like also need to move `require_downlevel_flags(VIEW_FORMATS)?` from `create_texture_view()` to `create_texture()`?

Comment (teoxoy) [reply wgpu-hal/src/vulkan/device.rs:556]: Oh yeah, ideally we'd check that as soon as possible. @cwfitzgerald is there any reason why we have it in `create_texture_view()` right now?

Regarding the new downlevel flag name, maybe we should shorten it to just `SURFACE_VIEW_FORMATS`?

Comment (jinleili) [reply wgpu-hal/src/vulkan/device.rs:556]: OK, renamed.

File: wgpu-core/src/device/mod.rs [hunk 4]
```diff
@@ -5303,7 +5303,7 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
         let (adapter_guard, mut token) = hub.adapters.read(&mut token);
         let (device_guard, _token) = hub.devices.read(&mut token);
 
-        let error = loop {
+        let error = 'outter: loop {
```
Comment (cwfitzgerald) [inline wgpu-core/src/device/mod.rs:5306]: Nit: 
```suggestion
        let error = 'outer: loop {
```

File: wgpu-hal/src/vulkan/device.rs (hunk 3 above)
Comment (cwfitzgerald) [reply wgpu-hal/src/vulkan/device.rs:556]: @teoxoy not that I know of - it should be texture creation

File: wgpu-hal/src/vulkan/adapter.rs [hunk 5]
```diff
@@ -320,7 +320,8 @@ impl PhysicalDeviceFeatures {
             | Df::UNRESTRICTED_INDEX_BUFFER
             | Df::INDIRECT_EXECUTION
             | Df::VIEW_FORMATS
-            | Df::UNRESTRICTED_EXTERNAL_TEXTURE_COPIES;
+            | Df::UNRESTRICTED_EXTERNAL_TEXTURE_COPIES
+            | Df::SURFACE_VIEW_FORMATS;
```
Comment (teoxoy) [inline wgpu-hal/src/vulkan/adapter.rs:324]: We need to set this downlevel flag based on support for `VK_KHR_swapchain_mutable_format`.

```rust
dl_flags.set(
    Df::SURFACE_VIEW_FORMATS,
    caps.supports_extension(vk::KhrSwapchainMutableFormatFn::name()),
);
```

File: wgpu-hal/src/vulkan/device.rs (hunk 3 above)
Comment (teoxoy) [inline wgpu-hal/src/vulkan/device.rs:556]: Now that we have the downlevel flag, we can remove this check.

**Review by teoxoy — CHANGES_REQUESTED** [review]
2 remaining things and we should be good to go.
Thanks for addressing all the comments!

**Comment by cwfitzgerald** [comment]
Adopting this PR so we can get it in this release

File: wgpu-hal/src/vulkan/adapter.rs (hunk 5 above)
Comment (teoxoy) [reply wgpu-hal/src/vulkan/adapter.rs:324]: @cwfitzgerald the flag still gets set here, even though it gets set/unset later, I think we should remove it for clarity.

Comment (cwfitzgerald) [reply wgpu-hal/src/vulkan/adapter.rs:324]: Oh whoops

**Pull request merged**